neo3/wallet/account.py

```diff
@@ -300,6 +300,51 @@ def sign(self, data: bytes, password: str) -> bytes:
     def create_new(cls, password: str) -> Account:
         return cls(password=password, watch_only=False)
 
+    def token_add(self, token_hash: types.UInt160) -> bool:
+        """
+        Add a script hash from a token in the extra property of this account.
```

Comment:
This description describes what the function internals do. I think we should at least (also?) describe the higher level purpose of the function. e.g.
```suggestion
        Associate a NEP17 token to the account. 
```
I'm open to alternative or more words. My line of thinking was; the function argument already describes that it is added based on the script hash so mentioning this feels like duplication to me. That the data is stored in the extra fields is an implementation detail, which might not be true in the future.

**Interface considerations**
The reason I specifically used "NEP17" in the description is because I think after this PR we should add some other helpers like `token_get_balance`. Otherwise, the add/delete functions have no real use-case. Currently, NEP17 is the only token standard for which we know how to query the chain for the balance, name etc. Which brings us to the fact that with the current interface we cannot validate that the script hash we're adding is actually a NEP17 token. Should we add a `snapshot` parameter such that we can query some chain data to see if it is NEP17? Or should we trust the user (:see_no_evil:) and in future functions like `token_get_balance` simply throw errors? (also pinging @lock9 for some extra UX input)

pro of not adding a snapshot argument:
- We don't couple account to chain data keeping it light. (The NEP17 validation could also be done via RPC.)

con of not adding a snapshot:
- A helper function like `token_transfer` will only fail on chain when we learn that the hash was not actually a NEP17 token. The user will also have payed GAS.

I'm leaning towards having a predicate as parameter that does the NEP17 validation. We can then provide a build in one based on a `snapshot`, and if people want to use another form (e.g. by RPC) they can create it themselves. Thoughts?

Reply:
I gave the interface some more thought and started thinking whether functions like `token_transfer` and `token_balance` should even be part of the wallet/account. To give an analogy; Your physical wallet is a container for holding your credit/debit cards. You can't ask the cards in your wallet what your balance is, nor to transfer anything. You ask an ATM machine for your balance and a payment terminal for transferring funds.

So..should we leave those functions out of the wallet/account class and into a utilities module perhaps?

Reply:
I understand the idea, but I think those methods should still be inside the wallet/account, mainly, because that's where they are in [Neo](https://github.com/neo-project/neo/blob/2040537350c889562085505f104479a25a5ec5aa/src/neo/Wallets/Wallet.cs#L241).

Reply:
Hi @ixje , I think that we could show a warning: "NEP-17 methods not found" or something like that. We only need the contract manifest to validate this, so either a snapshot, the manifest, or any kind of 'data provider' (as long as there is an interface to query for a contract manifest). 

I understand that the wallet is not where we store this kind of thing, but we need some class/object to be responsible for this role (managing invocations, like transfers). I think that it is common to have that added to the wallet. Maybe we could have a 'invocation manager' that could be responsible for that. At this moment, we need a way to query for the user balance, but in the future, there will be a need for other kinds of invoke, like querying for consensus candidates.

Maybe we could add, inside the wallet, or not, a class called Invocation Manager, and the wallet may use that to:
- Ensure if a contract has NEP-xx compatible methods
- Add an easy way to perform common invokes, like balanceOf, transfer etc.
 - I don't know if this class should only build the script, or if it should perform the full test-invoke. I think I prefer the second option.

I don't think it makes sense to have users depend on RPC calls if they are running a full node locally (even if they are querying themselves)

Reply:
While NEO has it in their wallet folder/package, they have a lot of tight coupling which I unfortunately had to copy in many places for time reasons. I think there is now room for other ways, see below. 

@lock9 Yeah I think having some helper module (e.g. InvocationManager) that can do the balance query etc makes sense. If it should really be in the `wallet` package...not entirely convinced yet. It's a contract related information query which makes me lean towards having the module under the `/contracts/` package. 

> I don't think it makes sense to have users depend on RPC calls if they are running a full node locally (even if they are querying themselves)
Remember, RPC is an option. Just because mamba has the capability of running as a full node, we're still an SDK first and should not enforce people to run full nodes. Of course if one can provide a `Snapshot` because they have a full node we should query that directly. But It should also be possible to query the data from else where.